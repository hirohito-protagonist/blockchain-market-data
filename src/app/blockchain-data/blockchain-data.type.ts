import {
  fromBlockchainDataMarketPriceEntity,
  fromBlockchainDataStatisticsEntity
} from './entities/index.entity';

export enum DataServiceType {
  Ticker = 'ticker',
  ToBTC = 'tobtc',
  Stats = 'stats'
}

export interface BTCQuery {
  value: number | string;
  currency: string;
}

export type QueryType = BTCQuery | null;

export type DataResponseType = fromBlockchainDataMarketPriceEntity.MarketPricesEntities |
fromBlockchainDataStatisticsEntity.StatisticEntity | number;
