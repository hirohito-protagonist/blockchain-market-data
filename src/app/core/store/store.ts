import {
  combineReducers,
  Action
} from '@ngrx/store';
import * as fromLayoutReducer from './layout.reducer';

export {
  fromLayoutReducer
};


export interface CoreState {
  layout: fromLayoutReducer.LayoutState;
  version: string;
}

export function featureVersion() {
  return '1.0.0';
}

export function reducers(state: CoreState | undefined, action: Action) {
  return combineReducers({
    version: featureVersion,
    layout: fromLayoutReducer.reducer
  })(state, action);
}
