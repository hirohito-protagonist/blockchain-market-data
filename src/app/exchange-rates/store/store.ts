import {
  combineReducers,
  Action,
  StoreModule
} from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';

import { Effects } from './effects';

import * as fromUIReducer from './ui.reducer';

export interface ExchangeRatesState {
  version: string;
  ui: fromUIReducer.UIState;
}

export function featureVersion() {
  return '1.0.0';
}

function reducers(state: ExchangeRatesState | undefined, action: Action) {
  return combineReducers({
    version: featureVersion,
    ui: fromUIReducer.reducer
  })(state, action);
}


export const Store = [
  StoreModule.forFeature('exchangeRates', reducers),
  EffectsModule.forFeature([Effects])
];
