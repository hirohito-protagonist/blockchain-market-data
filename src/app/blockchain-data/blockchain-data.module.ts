import { NgModule, ModuleWithProviders } from '@angular/core';

import { StoreModule } from '@ngrx/store';

import { reducers } from './reducers/index.reducer';

@NgModule({
  imports: [
    StoreModule.forFeature('blockchain-data', reducers)
  ]
})
export class BlockchainDataModule {

  static forRoot(): ModuleWithProviders {

    return {
      ngModule: BlockchainDataModule,
      providers: []
    };
  }
}
