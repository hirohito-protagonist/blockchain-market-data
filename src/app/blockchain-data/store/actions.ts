import { createAction, props } from '@ngrx/store';
import { DataServiceType, QueryType, DataResponseType } from './../blockchain-data.type';

export const fetchData = createAction(
  '[Blockchain Data Feature API] Fetch data',
  props<{ key: DataServiceType; query: QueryType; }>()
);

export const request = createAction(
  '[Blockchain Data Feature API] Data service request',
  props<{ key: DataServiceType; query: QueryType; }>()
);

export const response = createAction(
  '[Blockchain Data Feature API] Data service response',
  props<{ key: DataServiceType; response: DataResponseType; }>()
);
