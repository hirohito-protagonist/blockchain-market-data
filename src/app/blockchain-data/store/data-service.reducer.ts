import { createReducer, on } from '@ngrx/store';
import { request, response } from './../store/actions';
import { DataServiceType, QueryType, DataResponseType } from './../blockchain-data.type';

interface DataServiceNode<R, Q> {
  isFetching: boolean;
  response: R;
  lastUpdate: number;
  query: Q;
}

export interface DataServiceState {
  [key: string]: DataServiceNode<DataResponseType, QueryType>;
}

const initialState: DataServiceState = {
  [DataServiceType.Ticker]: {
    isFetching: false,
    response: null,
    lastUpdate: 0,
    query: null
  },
  [DataServiceType.ToBTC]: {
    isFetching: false,
    response: null,
    lastUpdate: 0,
    query: null
  },
  [DataServiceType.Stats]: {
    isFetching: false,
    response: null,
    lastUpdate: 0,
    query: null
  },
  [DataServiceType.Charts]: {
    isFetching: false,
    response: null,
    lastUpdate: 0,
    query: null
  }
};

export const reducer = createReducer(
  initialState,
  on(
    request,
    (state, { key, query }) => {

      if (state[key]) {

        return {
          ...state,
          [key]: {
            ...state[key],
            isFetching: true,
            response: null,
            query
          }
        };
      }
      return state;
    }
  ),
  on(
    response,
    (state, { key, response }) => {

      if (state[key]) {

        return {
          ...state,
          [key]: {
            ...state[key],
            isFetching: false,
            response,
            lastUpdate: Date.now()
          }
        };
      }
      return state;
    }
  )
);
