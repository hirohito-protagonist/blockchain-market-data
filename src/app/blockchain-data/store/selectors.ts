import { createFeatureSelector, createSelector } from '@ngrx/store';
import { DataServiceType } from './../blockchain-data.type';
import { DataServiceState } from './store';
import * as fromDataService from './data-service.reducer';

const getFeatureRootState = createFeatureSelector('blockchain-data');
const getDataServiceState = createSelector(getFeatureRootState, (s: DataServiceState) => s.data);
export const getServiceDataNode = (dataServiceType: DataServiceType) =>
  createSelector(getDataServiceState, (s: fromDataService.DataServiceState) => s[dataServiceType]);

export const lastUpdate = (dataServiceType: DataServiceType) =>
  createSelector(getServiceDataNode(dataServiceType), (node) => node.lastUpdate);

export const response = (dataServiceType: DataServiceType) =>
  createSelector(getServiceDataNode(dataServiceType), (node) => node.response);
