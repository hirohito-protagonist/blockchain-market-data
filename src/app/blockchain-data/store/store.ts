import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import * as fromDataService from './data-service.reducer';
import { DataServiceEffects } from './effects';

export interface DataServiceState {
  data: fromDataService.DataServiceState;
  version: string;
}

export function featureVersion() {
  return '1.0.0';
}

export const reducers = {
  data: fromDataService.reducer,
  version: featureVersion
};


export const Store = [
  StoreModule.forFeature('blockchain-data', reducers),
  EffectsModule.forFeature([ DataServiceEffects ])
];

