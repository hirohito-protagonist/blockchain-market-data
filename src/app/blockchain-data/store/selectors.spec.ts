import {
  getServiceDataNode,
  lastUpdate,
  response
} from './selectors';
import { reducers } from './store';
import { DataServiceType } from './../blockchain-data.type';


describe('Selectors', () => {

  let store;

  beforeEach(() => {

    store = {
      'blockchain-data': {
        data: reducers.data(void(0), {} as any),
        version: reducers.version()
      }
    };

  });

  it('should pick data service node', () => {

    expect(getServiceDataNode(DataServiceType.ToBTC)(store)).toMatchSnapshot();
  });

  it('should pick lastUpdate information for provided data service', () => {

    expect(lastUpdate(DataServiceType.ToBTC)(store)).toMatchSnapshot();
  });

  it('should pick response information for provided data service', () => {

    expect(response(DataServiceType.ToBTC)(store)).toMatchSnapshot();
  });
});
