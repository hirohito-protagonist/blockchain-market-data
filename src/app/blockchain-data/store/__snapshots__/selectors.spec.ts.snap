// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectors should pick data service node 1`] = `
Object {
  "isFetching": false,
  "lastUpdate": 0,
  "query": null,
  "response": null,
}
`;

exports[`Selectors should pick lastUpdate information for provided data service 1`] = `0`;

exports[`Selectors should pick response information for provided data service 1`] = `null`;
